### PR TITLE
Keepalive Timer

### PR DESCRIPTION
**Pull request by bokobza**
Added a timer to keep the node going until an interval has elapsed.
* This will be used by the Breeze and FullNode.UI apps in order to provide a redundancy scenario in case the UI does not successfully close the background full node process.
* A timer with a timeout of 60s is initialized in this manner: `dotnet run -testnet -keepalive=60`
* Every time a call is made to any API endpoint, the timer is reset.

### CODE REVIEW

File: src/Stratis.StratisD/Program.cs [hunk 1]
```diff
@@ -31,15 +33,22 @@ public static void Main(string[] args)
...
                 var node = new FullNodeBuilder()
                     .UseNodeSettings(nodeSettings)
                     .UseStratisConsensus()
                     .UseBlockStore()
                     .UseMempool()
                     .UseWallet()
                     .AddPowPosMining()
-                    .UseApi()
+                    .UseApi(null, options =>
+                    {
```
Comment (dangershony) [inline src/Stratis.StratisD/Program.cs]: Can we not move this in to the feature? would make the builder part much cleaner.

File: src/Stratis.Bitcoin.Features.Api/ApiFeature.cs [hunk 2]
```diff
@@ -58,7 +58,21 @@ public override void Initialize()
...
-            this.TryStartKeepaliveMonitor();
+            // Start the keepalive timer, if set.
+            // If the timer expires, the node will shut down.
+            if (this.apiFeatureOptions.KeepaliveTimer != null)
+            {
+                this.apiFeatureOptions.KeepaliveTimer.Elapsed += (sender, args) =>
+                {
+                    this.logger.LogInformation($"The application will shut down because the keepalive timer has elapsed.");
+
+                    this.apiFeatureOptions.KeepaliveTimer.Stop();
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Api/ApiFeature.cs]: I suggest adding this lines in the feature Dispose method as well.
```
  if (this.apiFeatureOptions.KeepaliveTimer != null)
  {
                   this.apiFeatureOptions.KeepaliveTimer.Stop();
                   this.apiFeatureOptions.KeepaliveTimer.Enabled = false;
```

Also can we call a dispose on the timer? (is the timer disposable?)

File: src/Stratis.Bitcoin.Features.Api/ApiFeature.cs [hunk 3]
```diff
@@ -74,41 +88,19 @@ public override void Dispose()
...
-        public KeepaliveMonitor KeepaliveMonitor { get; private set; }
+        public Timer KeepaliveTimer { get; private set; }
 
         public void Keepalive(TimeSpan timeSpan)
         {
-            this.KeepaliveMonitor = new KeepaliveMonitor { KeepaliveInterval = timeSpan };
+            this.KeepaliveTimer = new Timer
+            {
+                AutoReset = false,
+                Interval = timeSpan.TotalSeconds * 1000
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.Api/ApiFeature.cs]: Why the multiplication?
`timeSpan.TotalSeconds * 1000`

**Comment by dangershony** [comment]
Minor comments looks good.

File: src/Stratis.Bitcoin.Features.Api/ApiFeature.cs (hunk 3 above)
Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Api/ApiFeature.cs]: because the interval takes in miliseconds.

File: src/Stratis.StratisD/Program.cs (hunk 1 above)
Comment (bokobza) [reply src/Stratis.StratisD/Program.cs]: no you can't that's how this is designed. You get an options object, that you need to initialize here.

File: src/Stratis.Bitcoin.Features.Api/ApiFeature.cs (hunk 2 above)
Comment (bokobza) [reply src/Stratis.Bitcoin.Features.Api/ApiFeature.cs]: good idea :-)

File: src/Stratis.StratisD/Program.cs (hunk 1 above)
Comment (dangershony) [reply src/Stratis.StratisD/Program.cs]: Fair enough. 
Optional we could add a second method in the feature to do that.
```
 .UseKeepAliveApi(args)
```

File: src/Stratis.Bitcoin.Features.Api/ApiFeature.cs (hunk 3 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.Api/ApiFeature.cs]: Ah in that case rename the method parameter to `keepaliveSeconds` thats how we do in other places.

File: src/Stratis.StratisD/Program.cs (hunk 1 above)
Comment (bokobza) [reply src/Stratis.StratisD/Program.cs]: Don't really like passing the whole args objects to features.
Also methods that look like  .UseXXX() are really used for features and I think the keepalive should stay as an api option as 1. it depends on it 2. it really means that if the api hasn't been called, kill the app.

**Pull request merged**